### PR TITLE
Quasi-random Fourier Features

### PR DESCRIPTION
**Pull request by ltiao**

### CODE REVIEW

**Comment by vdutor** [comment]
Hi @ltiao, would you mind updating the base branch so that I only see the relevant changes?

File: gpflux/layers/basis_functions/fourier_features/random.py [hunk 1]
```diff
@@ -122,6 +123,43 @@ def _compute_constant(self) -> tf.Tensor:
         return self.rff_constant(self.kernel.variance, output_dim=2 * self.n_components)
 
 
+class QuasiRandomFourierFeatures(RandomFourierFeatures):
+
+    r"""
+    Quasi-random Fourier features (ORF) :cite:p:`yang2014quasi` for more
```
Comment (vdutor) [inline gpflux/layers/basis_functions/fourier_features/random.py]: Looks like the acronym is wrong?
```suggestion
    Quasi-random Fourier features (???) :cite:p:`yang2014quasi` for more
```

File: gpflux/layers/basis_functions/fourier_features/random.py [hunk 2]
```diff
@@ -122,6 +123,43 @@ def _compute_constant(self) -> tf.Tensor:
...
+    ENGINE_CLASSES = dict(sobol=Sobol, halton=Halton)
+
+    def __init__(
+        self,
+        kernel: gpflow.kernels.Kernel,
+        n_components: int,
+        method: str = "halton",
+        scramble: bool = True,
+        **kwargs: Mapping
+    ):
```
Comment (vdutor) [inline gpflux/layers/basis_functions/fourier_features/random.py]: Please add docstring

File: gpflux/layers/basis_functions/fourier_features/random.py [hunk 3]
```diff
@@ -122,6 +123,43 @@ def _compute_constant(self) -> tf.Tensor:
...
+        assert isinstance(kernel, gpflow.kernels.SquaredExponential), "Unsupported Kernel"
```
Comment (vdutor) [inline gpflux/layers/basis_functions/fourier_features/random.py]: To be consistent you probably want to have a `SUPPORTED_KERNELS` variable as well

File: gpflux/layers/basis_functions/fourier_features/random.py [hunk 4]
```diff
@@ -122,6 +123,43 @@ def _compute_constant(self) -> tf.Tensor:
...
+        assert method in self.ENGINE_CLASSES, "Quasi Monte Carlo method not recognized"
```
Comment (vdutor) [inline gpflux/layers/basis_functions/fourier_features/random.py]: Maybe tell the user what the options are.

File: gpflux/layers/basis_functions/fourier_features/random.py [hunk 5]
```diff
@@ -122,6 +123,43 @@ def _compute_constant(self) -> tf.Tensor:
...
+        self.engine_cls = self.ENGINE_CLASSES[method]
+        self.scramble = scramble
+
+    def _weights_init(self, shape: TensorType, dtype: Optional[DType] = None) -> TensorType:
+        n_components, input_dim = shape  # M, D
+        if input_dim > 1:
+            engine = self.engine_cls(d=input_dim, scramble=self.scramble)
+        else:
+            # When `input_dim==1`, specifying an engine will erroneously throw
+            # exception.
```
Comment (vdutor) [inline gpflux/layers/basis_functions/fourier_features/random.py]:
```suggestion
            # an exception.
```

File: gpflux/layers/basis_functions/fourier_features/random.py [hunk 6]
```diff
@@ -122,6 +123,43 @@ def _compute_constant(self) -> tf.Tensor:
...
+            # Can remove once issue https://github.com/scipy/scipy/issues/14904 is resolved.
+            warnings.warn("QMC engine option ignored when `input_dim==1` (defaults to `Sobol`)")
+            engine = None
```
Comment (vdutor) [inline gpflux/layers/basis_functions/fourier_features/random.py]: Does it also crash if you set this one to `Sobol`?

File: tests/gpflux/layers/basis_functions/fourier_features/test_random.py [hunk 7]
```diff
@@ -71,7 +82,7 @@ def test_throw_for_unsupported_kernel(basis_func_cls):
 
 
 def test_random_fourier_features_can_approximate_kernel_multidim(
-    basis_func_cls, kernel_cls, variance, lengthscale, n_dims
+    random_basis_func_cls, kernel_cls, variance, lengthscale, n_dims
 ):
     n_components = 40000
```
Comment (vdutor) [inline tests/gpflux/layers/basis_functions/fourier_features/test_random.py]: how did you set this value?

File: tests/gpflux/layers/basis_functions/fourier_features/test_random.py [hunk 8]
```diff
@@ -81,7 +92,32 @@ def test_random_fourier_features_can_approximate_kernel_multidim(
...
+
+    actual_kernel_matrix = kernel.K(x, y)
+
+    np.testing.assert_allclose(approx_kernel_matrix, actual_kernel_matrix, atol=5e-2)
+
+
+def test_quasi_random_fourier_features_can_approximate_kernel_multidim(
+    variance, lengthscale, n_dims
+):
+    n_components = 25000
```
Comment (vdutor) [inline tests/gpflux/layers/basis_functions/fourier_features/test_random.py:117]: how did you set this value?

**Review by vdutor — APPROVED** [review]

File: gpflux/layers/basis_functions/fourier_features/random.py (hunk 6 above)
Comment (ltiao) [reply gpflux/layers/basis_functions/fourier_features/random.py]: Yes, apparently it's deliberate https://github.com/scipy/scipy/issues/14904#issuecomment-950381060

**Pull request closed without merging**